In [5]:
import pandas as pd
import numpy as np
import datetime

In [6]:
stalk_data = pd.read_pickle('Stalk_Cleanest_v3.pickle')
stalk_data.head()

User_ID           Date_Time  Location_ID
0        0 2010-10-18 22:17:43       420315
1        0 2010-10-17 23:42:03       316637
2        0 2010-10-17 19:26:05        16516
3        0 2010-10-16 18:50:42      5535878
4        0 2010-10-12 23:58:03        15372

### Dataframe sorted based on Location_ID and Date_Time for preliminary scanning of apparent patterns.

In [7]:
sorted_stalk_data = stalk_data.sort_values(by=['Location_ID', 'Date_Time'])
sorted_stalk_data.head(25)

User_ID           Date_Time  Location_ID
6281          24 2009-02-05 06:27:43         8904
101570       343 2009-03-08 02:09:11         8904
101539       343 2009-05-28 23:40:31         8904
76715        256 2009-05-30 03:33:23         8904
101530       343 2009-06-20 01:40:01         8904
101511       343 2009-08-21 03:01:19         8904
189075       576 2009-10-05 20:19:24         8904
88182        310 2009-12-06 22:51:49         8904
117500       392 2009-12-17 23:02:42         8904
1123876     5164 2010-01-25 01:29:46         8904
76404        256 2010-04-24 23:57:00         8904
76194        256 2010-09-19 01:47:05         8904
2741249    32391 2009-05-27 20:59:09         8932
105493       350 2009-08-16 21:48:27         8932
1226028     6046 2010-02-20 01:08:20         8932
1000650     4648 2010-02-21 01:17:07         8932
4730125   105294 2010-03-17 16:42:29         8932
3309706    44458 2010-04-20 17:56:37         8932
1008534     4678 2010-05-21 22:53:16         8932
1000392     4648 2010-05-23 22:34:42         8932
3301718    44297 2010-06-18 23:54:11         8932
1438375     9565 2010-06-29 23:21:12         8932
3309664    44458 2010-07-25 18:13:48         8932
984240      4571 2010-07-29 23:17:36         8932
3876753    63756 2010-08-02 01:17:30         8932

### Create separate Date and Time columns for easier manipulation.

In [8]:
rm_time_stalk_data = sorted_stalk_data
rm_time_stalk_data['Date'] = [d.date() for d in rm_time_stalk_data['Date_Time']]
rm_time_stalk_data['Time'] = [d.time() for d in rm_time_stalk_data['Date_Time']]
rm_time_stalk_data = rm_time_stalk_data.drop('Date_Time', axis=1)
rm_time_stalk_data.head(25)

User_ID  Location_ID        Date      Time
6281          24         8904  2009-02-05  06:27:43
101570       343         8904  2009-03-08  02:09:11
101539       343         8904  2009-05-28  23:40:31
76715        256         8904  2009-05-30  03:33:23
101530       343         8904  2009-06-20  01:40:01
101511       343         8904  2009-08-21  03:01:19
189075       576         8904  2009-10-05  20:19:24
88182        310         8904  2009-12-06  22:51:49
117500       392         8904  2009-12-17  23:02:42
1123876     5164         8904  2010-01-25  01:29:46
76404        256         8904  2010-04-24  23:57:00
76194        256         8904  2010-09-19  01:47:05
2741249    32391         8932  2009-05-27  20:59:09
105493       350         8932  2009-08-16  21:48:27
1226028     6046         8932  2010-02-20  01:08:20
1000650     4648         8932  2010-02-21  01:17:07
4730125   105294         8932  2010-03-17  16:42:29
3309706    44458         8932  2010-04-20  17:56:37
1008534     4678         8932  2010-05-21  22:53:16
1000392     4648         8932  2010-05-23  22:34:42
3301718    44297         8932  2010-06-18  23:54:11
1438375     9565         8932  2010-06-29  23:21:12
3309664    44458         8932  2010-07-25  18:13:48
984240      4571         8932  2010-07-29  23:17:36
3876753    63756         8932  2010-08-02  01:17:30

### Selecting duplicated values of Date and Location_ID will further narrow down the possibilities:

In [9]:
date_time_n_locs = rm_time_stalk_data[['Date', 'Location_ID']]
dup_time_stalk_data = rm_time_stalk_data[date_time_n_locs.duplicated(keep=False)]
dup_time_stalk_data.tail(25)

User_ID  Location_ID        Date      Time
1756699    12947      5868441  2010-10-21  15:37:38
6347234   183758      5882780  2010-10-21  14:01:54
6347232   183758      5882780  2010-10-21  18:51:51
6347229   183758      5882780  2010-10-22  10:51:28
6347223   183758      5882780  2010-10-22  21:13:42
6347233   183758      5899658  2010-10-21  18:16:56
6347231   183758      5899658  2010-10-21  21:24:17
6370270   186903      5901335  2010-10-22  16:28:15
6379154   187598      5901335  2010-10-22  19:29:00
4440991    98061      5904186  2010-10-21  19:28:44
2313447    20628      5904186  2010-10-21  19:41:04
1418356     9046      5909364  2010-10-21  22:06:36
502706      1921      5909364  2010-10-21  23:47:01
484567      1773      5912633  2010-10-22  00:48:02
264929       842      5912633  2010-10-22  01:28:22
185667       573      5912882  2010-10-21  23:31:53
257800       819      5912882  2010-10-21  23:33:36
6357774   185298      5914456  2010-10-22  00:34:51
6357771   185298      5914456  2010-10-22  21:42:32
1800224    13943      5927439  2010-10-22  10:15:55
3884788    64168      5927439  2010-10-22  10:57:28
6256184   175543      5928544  2010-10-22  12:11:37
6256183   175543      5928544  2010-10-22  20:40:29
481607      1762      5930620  2010-10-22  13:17:47
1440832     9619      5930620  2010-10-22  13:58:44

#### Since two rows constitutes a pair of users, transforming it into a single row will make analysis more convenient. We will each user pair in two separate dataframes. We will only use User_ID and Time to create our second dataframe since rest of the values are same for both users.

In [10]:
row_num = 0
dup_time_stalk_data_friend_dictionary = {
    'User_ID2': [],
    'Time2': []
}
indices_to_drop = []
for i, row in dup_time_stalk_data.iterrows():
    row_num += 1
    if row_num%2==0:
        dup_time_stalk_data_friend_dictionary['User_ID2'].append(row['User_ID'])
        dup_time_stalk_data_friend_dictionary['Time2'].append(row['Time'])
        indices_to_drop.append(i)
dup_time_stalk_data1 = dup_time_stalk_data.drop(indices_to_drop, axis=0)
dup_time_stalk_data1 = dup_time_stalk_data1.set_index(np.arange(dup_time_stalk_data1.shape[0]))
dup_time_stalk_data2 = pd.DataFrame(dup_time_stalk_data_friend_dictionary)

In [11]:
dup_time_stalk_data1.head(25)

User_ID  Location_ID        Date      Time
0     63756         8932  2010-08-02  01:17:30
1     28882         8938  2009-09-27  01:33:43
2     28882         8938  2009-10-24  00:40:02
3     28882         8938  2009-11-16  16:03:41
4     36254         8938  2009-11-24  13:42:47
5       488         8938  2009-11-24  18:24:05
6       256         8938  2009-12-12  21:47:48
7      5164         8938  2009-12-14  00:33:38
8       107         8938  2009-12-20  21:21:13
9       256         8938  2009-12-23  19:16:08
10      256         8938  2009-12-30  20:21:43
11    36254         8938  2010-02-04  22:37:43
12    28882         8938  2010-02-06  21:22:27
13      570         8938  2010-02-18  14:23:17
14      488         8938  2010-02-27  23:33:57
15      488         8938  2010-03-19  17:56:26
16    28882         8938  2010-03-28  22:55:58
17      488         8938  2010-04-02  16:37:02
18      488         8938  2010-04-23  16:58:54
19   114144         8938  2010-05-16  19:18:56
20      488         8938  2010-09-14  13:29:16
21      602         8947  2009-08-11  05:46:51
22      602         8947  2009-08-11  21:53:39
23     4936         8947  2009-10-07  20:14:32
24     5838         8947  2009-10-17  03:55:44

In [12]:
dup_time_stalk_data2.head(25)

Time2  User_ID2
0   01:32:38      2356
1   23:08:54       576
2   22:29:05      8616
3   16:20:53     24332
4   16:33:52     28882
5   20:37:15      4893
6   00:28:35      4893
7   02:52:33       256
8   16:08:59       570
9   17:40:46       558
10  18:52:36       488
11  18:40:45      4207
12  00:26:04       488
13  19:58:44       488
14  15:08:39       256
15  14:40:02       570
16  14:28:53     36254
17  16:57:24       273
18  14:46:25       488
19  19:19:42     39765
20  19:10:51       488
21  18:00:29       602
22  01:13:18    113913
23  00:02:40      5831
24  00:45:55       518

## Combine both dataframes:

In [13]:
dup_time_stalk_data_transformed = pd.concat([dup_time_stalk_data1, dup_time_stalk_data2], axis=1)
dup_time_stalk_data_transformed.tail()

User_ID  Location_ID        Date      Time     Time2  User_ID2
684558      819      5912882  2010-10-21  23:33:36  00:34:51  185298.0
684559   185298      5914456  2010-10-22  21:42:32  10:15:55   13943.0
684560    64168      5927439  2010-10-22  10:57:28  12:11:37  175543.0
684561   175543      5928544  2010-10-22  20:40:29  13:17:47    1762.0
684562     9619      5930620  2010-10-22  13:58:44       NaN       NaN

## Debug missing values:

In [14]:
row_num = 0
dup_time_stalk_data_shape = dup_time_stalk_data.shape[0]# For debugging
dup_time_stalk_data_friend_dictionary = {
    'User_ID2': [],
    'Time2': []
}
indices_to_drop = []
for i, row in dup_time_stalk_data.iterrows():
    row_num += 1
    if row_num%2==0:
        dup_time_stalk_data_friend_dictionary['User_ID2'].append(row['User_ID'])
        dup_time_stalk_data_friend_dictionary['Time2'].append(row['Time'])
        indices_to_drop.append(i)
dup_time_stalk_data1 = dup_time_stalk_data.drop(indices_to_drop, axis=0)
dup_time_stalk_data1 = dup_time_stalk_data1.set_index(np.arange(dup_time_stalk_data1.shape[0]))
dup_time_stalk_data2 = pd.DataFrame(dup_time_stalk_data_friend_dictionary)

Number of rows in dup_time_stalk_data_shape:

In [15]:
dup_time_stalk_data_shape

1369125

The expected number of rows should be an even number (user pairs). The odd number of rows could be because:
    1. A user visited a location more than once a day.
    2. More than two users could have visited the same location on the same day.
To verify this, add columns Location_ID2 and Date2:

In [16]:
row_num = 0
dup_time_stalk_data_friend_dictionary = {
    'User_ID2': [],
    'Location_ID2': [],
    'Date2': [],
    'Time2': []
}
indices_to_drop = []
for i, row in dup_time_stalk_data.iterrows():
    row_num += 1
    if row_num%2==0:
        dup_time_stalk_data_friend_dictionary['User_ID2'].append(row['User_ID'])
        dup_time_stalk_data_friend_dictionary['Location_ID2'].append(row['Location_ID'])
        dup_time_stalk_data_friend_dictionary['Date2'].append(row['Date'])
        dup_time_stalk_data_friend_dictionary['Time2'].append(row['Time'])
        indices_to_drop.append(i)
dup_time_stalk_data1 = dup_time_stalk_data.drop(indices_to_drop, axis=0)
dup_time_stalk_data1 = dup_time_stalk_data1.set_index(np.arange(dup_time_stalk_data1.shape[0]))
dup_time_stalk_data2 = pd.DataFrame(dup_time_stalk_data_friend_dictionary)

In [17]:
dup_time_stalk_data_transformed = pd.concat([dup_time_stalk_data1, dup_time_stalk_data2], axis=1)
dup_time_stalk_data_transformed.tail()

User_ID  Location_ID        Date      Time       Date2  Location_ID2  \
684558      819      5912882  2010-10-21  23:33:36  2010-10-22     5914456.0   
684559   185298      5914456  2010-10-22  21:42:32  2010-10-22     5927439.0   
684560    64168      5927439  2010-10-22  10:57:28  2010-10-22     5928544.0   
684561   175543      5928544  2010-10-22  20:40:29  2010-10-22     5930620.0   
684562     9619      5930620  2010-10-22  13:58:44         NaN           NaN   

           Time2  User_ID2  
684558  00:34:51  185298.0  
684559  10:15:55   13943.0  
684560  12:11:37  175543.0  
684561  13:17:47    1762.0  
684562       NaN       NaN

### The above error in hypothesis could have skewed our grouping of user pairs. 

In [18]:
prev_row = None
prev_i = None
activate = False
indices_to_drop = []
for i, row in dup_time_stalk_data[['User_ID', 'Location_ID', 'Date']].iterrows():
    current_row = row
    current_i = i
    
    if activate:
        Filter = (current_row['User_ID'] == prev_row['User_ID']) & \
        (current_row['Location_ID'] == prev_row['Location_ID']) & \
        (current_row['Date'] == prev_row['Date'])
    
        if Filter:
            indices_to_drop.append(prev_i)
            indices_to_drop.append(current_i)
    prev_row = current_row
    prev_i = current_i
    activate = True
dup_time_stalk_data.loc[indices_to_drop,:].head(25)

User_ID  Location_ID        Date      Time
153629       488         8938  2010-02-27  19:58:44
153628       488         8938  2010-02-27  23:33:57
153446       488         8938  2010-09-14  13:29:16
153445       488         8938  2010-09-14  19:10:51
199359       602         8947  2009-08-11  05:46:51
199358       602         8947  2009-08-11  18:00:29
199358       602         8947  2009-08-11  18:00:29
199356       602         8947  2009-08-11  21:53:39
164981       518         8947  2009-10-22  00:45:55
164980       518         8947  2009-10-22  18:14:41
164964       518         8947  2009-11-05  01:33:27
164962       518         8947  2009-11-05  19:11:07
153399       487         8947  2009-11-16  17:24:47
153397       487         8947  2009-11-16  21:24:52
158018       494         8947  2009-12-16  21:31:16
158017       494         8947  2009-12-16  21:31:36
165095       519         8947  2010-02-26  03:58:21
165094       519         8947  2010-02-26  16:19:58
2054252    18315         8947  2010-02-26  16:33:39
2054250    18315         8947  2010-02-26  22:53:36
199344       602         8947  2010-02-27  04:31:36
199340       602         8947  2010-02-27  20:53:08
164863       518         8947  2010-03-11  18:46:56
164861       518         8947  2010-03-11  21:52:38
140509       457         8947  2010-03-26  05:27:21

### Remove all rows where a user has visited a location again on the same day.

In [19]:
prev_row = None
activate = False
indices_to_drop = []
for i, row in dup_time_stalk_data[['User_ID', 'Location_ID', 'Date']].iterrows():
    current_row = row
    current_i = i
    
    if activate:
        Filter = (current_row['User_ID'] == prev_row['User_ID']) & \
        (current_row['Location_ID'] == prev_row['Location_ID']) & \
        (current_row['Date'] == prev_row['Date'])
    
        if Filter:
            indices_to_drop.append(current_i)
    prev_row = current_row
    activate = True
dup_time_stalk_data = dup_time_stalk_data.drop(indices_to_drop, axis=0)

In [20]:
dup_time_stalk_data.head(25)

User_ID  Location_ID        Date      Time
3876753    63756         8932  2010-08-02  01:17:30
573886      2356         8932  2010-08-02  01:32:38
2615426    28882         8938  2009-09-27  01:33:43
189340       576         8938  2009-09-27  23:08:54
2615381    28882         8938  2009-10-24  00:40:02
1390478     8616         8938  2009-10-24  22:29:05
2615349    28882         8938  2009-11-16  16:03:41
2435622    24332         8938  2009-11-16  16:20:53
2910831    36254         8938  2009-11-24  13:42:47
2615337    28882         8938  2009-11-24  16:33:52
153733       488         8938  2009-11-24  18:24:05
1075933     4893         8938  2009-12-12  20:37:15
76631        256         8938  2009-12-12  21:47:48
1075919     4893         8938  2009-12-14  00:28:35
1124223     5164         8938  2009-12-14  00:33:38
76618        256         8938  2009-12-20  02:52:33
21311        107         8938  2009-12-20  21:21:13
185655       570         8938  2009-12-23  16:08:59
76614        256         8938  2009-12-23  19:16:08
182563       558         8938  2009-12-30  17:40:46
76602        256         8938  2009-12-30  20:21:43
153643       488         8938  2010-02-04  18:52:36
2910819    36254         8938  2010-02-04  22:37:43
925992      4207         8938  2010-02-06  18:40:45
2615278    28882         8938  2010-02-06  21:22:27

In [21]:
dup_time_stalk_data.shape[0]

1212019

#### The number of rows is still an odd number. This confirms our error in hypothesis as the indices 291083, 2615337, 153733 share the same date and location.

In [22]:
dup_time_stalk_data.loc[[2910831, 2615337, 153733], :]

User_ID  Location_ID        Date      Time
2910831    36254         8938  2009-11-24  13:42:47
2615337    28882         8938  2009-11-24  16:33:52
153733       488         8938  2009-11-24  18:24:05

#### Solution is to drop one user. Since we are trying to find the closest user pairs. It is logical to drop the user with the highest time difference between each other:

In [23]:
prev_row = None
activate = False
indices_to_drop = []
temp_indices = []
temp_times = []
prev_i = None

def time_to_seconds(t):
    time_delta = datetime.timedelta(hours=t.hour, minutes=t.minute, seconds=t.second)
    return int(time_delta.total_seconds())

for i, row in dup_time_stalk_data[['Location_ID', 'Date', 'Time']].iterrows():
    current_row = row
    current_i = i
    
    if activate:
        Filter = (current_row['Location_ID'] == prev_row['Location_ID']) & \
        (current_row['Date'] == prev_row['Date'])
    
        if Filter:
            if prev_i not in temp_indices:
                temp_indices.append(prev_i)
                temp_times.append(time_to_seconds(dup_time_stalk_data.loc[prev_i, 'Time']))
            if current_i not in temp_indices:
                temp_indices.append(current_i)
                temp_times.append(time_to_seconds(dup_time_stalk_data.loc[current_i, 'Time']))
        else:
            if len(temp_times) > 2:
                while len(temp_times) > 2:
                    sub_result = []
                    for li in range(len(temp_times)):
                        sub_result.append(abs(temp_times[li] - np.mean(temp_times)))
                    lst_index_of_max_diff = sub_result.index(max(sub_result))
                    indices_to_drop.append(temp_indices[lst_index_of_max_diff])
                    del temp_times[lst_index_of_max_diff]
                    del temp_indices[lst_index_of_max_diff]
            temp_times = []
            temp_indices = []
    prev_row = current_row
    prev_i = current_i
    activate = True
dup_time_stalk_data_keep_two = dup_time_stalk_data.drop(indices_to_drop, axis=0)

In [24]:
dup_time_stalk_data_keep_two.head(25)

User_ID  Location_ID        Date      Time
3876753    63756         8932  2010-08-02  01:17:30
573886      2356         8932  2010-08-02  01:32:38
2615426    28882         8938  2009-09-27  01:33:43
189340       576         8938  2009-09-27  23:08:54
2615381    28882         8938  2009-10-24  00:40:02
1390478     8616         8938  2009-10-24  22:29:05
2615349    28882         8938  2009-11-16  16:03:41
2435622    24332         8938  2009-11-16  16:20:53
2615337    28882         8938  2009-11-24  16:33:52
153733       488         8938  2009-11-24  18:24:05
1075933     4893         8938  2009-12-12  20:37:15
76631        256         8938  2009-12-12  21:47:48
1075919     4893         8938  2009-12-14  00:28:35
1124223     5164         8938  2009-12-14  00:33:38
76618        256         8938  2009-12-20  02:52:33
21311        107         8938  2009-12-20  21:21:13
185655       570         8938  2009-12-23  16:08:59
76614        256         8938  2009-12-23  19:16:08
182563       558         8938  2009-12-30  17:40:46
76602        256         8938  2009-12-30  20:21:43
153643       488         8938  2010-02-04  18:52:36
2910819    36254         8938  2010-02-04  22:37:43
925992      4207         8938  2010-02-06  18:40:45
2615278    28882         8938  2010-02-06  21:22:27
153634       488         8938  2010-02-18  00:26:04

In [25]:
dup_time_stalk_data_keep_two.shape[0]

948636

### Rows are an even number. Create new dataframe containing user pairs in each row:

In [26]:
row_num = 0
dup_time_stalk_data_keep_two_friend_dictionary = {
    'User_ID2': [],
    'Time2': []
}
indices_to_drop = []
for i, row in dup_time_stalk_data_keep_two.iterrows():
    row_num += 1
    if row_num%2==0:
        dup_time_stalk_data_keep_two_friend_dictionary['User_ID2'].append(row['User_ID'])
        dup_time_stalk_data_keep_two_friend_dictionary['Time2'].append(row['Time'])
        indices_to_drop.append(i)
dup_time_stalk_data_keep_two1 = dup_time_stalk_data_keep_two.drop(indices_to_drop, axis=0)

dup_time_stalk_data_keep_two1 = dup_time_stalk_data_keep_two1.\
    set_index(np.arange(dup_time_stalk_data_keep_two1.shape[0]))
    
dup_time_stalk_data_keep_two2 = pd.DataFrame(dup_time_stalk_data_keep_two_friend_dictionary)

dup_time_stalk_data_keep_two_transformed = pd.concat([dup_time_stalk_data_keep_two1, 
                                                      dup_time_stalk_data_keep_two2], axis=1)
dup_time_stalk_data_keep_two_transformed.tail()

User_ID  Location_ID        Date      Time     Time2  User_ID2
474313     1773      5912633  2010-10-22  00:48:02  01:28:22       842
474314      573      5912882  2010-10-21  23:31:53  23:33:36       819
474315   185298      5914456  2010-10-22  00:34:51  10:15:55     13943
474316    64168      5927439  2010-10-22  10:57:28  12:11:37    175543
474317     1762      5930620  2010-10-22  13:17:47  13:58:44      9619

## Check tail to verify:

In [27]:
dup_time_stalk_data.tail()

User_ID  Location_ID        Date      Time
1800224    13943      5927439  2010-10-22  10:15:55
3884788    64168      5927439  2010-10-22  10:57:28
6256184   175543      5928544  2010-10-22  12:11:37
481607      1762      5930620  2010-10-22  13:17:47
1440832     9619      5930620  2010-10-22  13:58:44

## Find closest pair in terms of seconds:

In [28]:
dup_time_stalk_data_keep_two_transformed['Score'] = \
abs(dup_time_stalk_data_keep_two_transformed['Time2'].apply(time_to_seconds) - \
dup_time_stalk_data_keep_two_transformed['Time'].apply(time_to_seconds))
dup_time_stalk_data_keep_two_transformed = dup_time_stalk_data_keep_two_transformed.sort_values('Score')
dup_time_stalk_data_keep_two_transformed.head(25)

User_ID  Location_ID        Date      Time     Time2  User_ID2  Score
360413    15389       790490  2010-05-01  18:22:42  18:22:42     99873      0
357797   103754       775389  2010-09-26  15:38:29  15:38:29    103755      0
260743    12877       280992  2010-08-03  20:58:05  20:58:05     94240      0
297026     2800       424351  2010-10-03  22:39:01  22:39:01      3353      0
445572    15466      1603426  2010-08-16  18:57:13  18:57:13     43153      0
350468    51390       727538  2010-08-17  23:07:21  23:07:21     55371      0
17798       104        10798  2010-01-30  23:11:01  23:11:01      8555      0
69013       189        25586  2010-06-18  23:54:42  23:54:42       190      0
75451       947        27886  2009-11-05  19:56:51  19:56:51      3935      0
338883     1352       664108  2010-07-31  19:17:08  19:17:08      1353      0
110731      750        48975  2010-10-19  23:05:08  23:05:08     10837      0
388983     3121      1014950  2010-10-02  15:23:30  15:23:30      3440      0
5966       2861         9313  2010-08-15  03:50:00  03:50:00     38065      0
69090      6549        25628  2010-03-10  03:39:40  03:39:40      7867      0
175116    19941        95443  2010-05-29  18:49:55  18:49:55     40677      0
208776     5611       140002  2010-04-10  05:47:16  05:47:16      9204      0
193380    38701       114843  2010-07-14  19:35:26  19:35:26     38907      0
350413    22449       727296  2010-06-21  11:37:11  11:37:11     83954      0
264065      456       293641  2010-09-24  23:48:33  23:48:33      4216      0
347533     5689       711185  2010-04-10  20:58:31  20:58:31      5717      0
383829    39367       971523  2010-06-23  20:17:49  20:17:49     79484      0
354704    26612       755984  2010-09-30  23:48:50  23:48:50    119031      0
298445    19935       431723  2010-01-22  19:21:42  19:21:42    109724      0
136582      573        67622  2010-03-20  14:03:57  14:03:57       573      0
112202      562        50121  2010-09-11  15:28:16  15:28:16     36070      0

#### It seems there are many user pairs who arrive at the same location at the exact same second. In order to determine a 'Friend Pair' and 'Non-Friend Pair' we might need to generate a scoring system based on the frequency of meetups. In order to narrow things down, let's remove pairs that are not duplicated.

In [29]:
user_pair = dup_time_stalk_data_keep_two_transformed[['User_ID', 'User_ID2']]
dup_user_pairs = dup_time_stalk_data_keep_two_transformed[user_pair.duplicated(keep=False)]
dup_user_pairs.head(25)

User_ID  Location_ID        Date      Time     Time2  User_ID2  Score
360413    15389       790490  2010-05-01  18:22:42  18:22:42     99873      0
357797   103754       775389  2010-09-26  15:38:29  15:38:29    103755      0
260743    12877       280992  2010-08-03  20:58:05  20:58:05     94240      0
297026     2800       424351  2010-10-03  22:39:01  22:39:01      3353      0
445572    15466      1603426  2010-08-16  18:57:13  18:57:13     43153      0
350468    51390       727538  2010-08-17  23:07:21  23:07:21     55371      0
69013       189        25586  2010-06-18  23:54:42  23:54:42       190      0
75451       947        27886  2009-11-05  19:56:51  19:56:51      3935      0
338883     1352       664108  2010-07-31  19:17:08  19:17:08      1353      0
110731      750        48975  2010-10-19  23:05:08  23:05:08     10837      0
388983     3121      1014950  2010-10-02  15:23:30  15:23:30      3440      0
175116    19941        95443  2010-05-29  18:49:55  18:49:55     40677      0
208776     5611       140002  2010-04-10  05:47:16  05:47:16      9204      0
350413    22449       727296  2010-06-21  11:37:11  11:37:11     83954      0
264065      456       293641  2010-09-24  23:48:33  23:48:33      4216      0
347533     5689       711185  2010-04-10  20:58:31  20:58:31      5717      0
383829    39367       971523  2010-06-23  20:17:49  20:17:49     79484      0
354704    26612       755984  2010-09-30  23:48:50  23:48:50    119031      0
298445    19935       431723  2010-01-22  19:21:42  19:21:42    109724      0
136582      573        67622  2010-03-20  14:03:57  14:03:57       573      0
112202      562        50121  2010-09-11  15:28:16  15:28:16     36070      0
11361      1051         9948  2010-06-12  04:52:16  04:52:16     10410      0
432106    10147      1443730  2010-07-15  15:44:02  15:44:02     88102      0
444022     1559      1584827  2010-10-19  08:48:20  08:48:20     53283      0
233934    15281       195809  2009-12-23  18:29:28  18:29:28     99636      0

#### Let's create a new DataFrame which contains the User_Pair, Meet_Count, Average_Seconds and Score columns. User_Pair will be the concatenation of User_ID and User_ID2 sepearated by a comma. Meet_Count counts the total number of meetings and Score will contain the average seconds of the User_Pair with a weighted quotient. This quotient is the average score divided by Meet_Count. This will reveal and prioritize friends who meet each other more often.

In [30]:
user_pair_scores = dup_user_pairs.drop(['Location_ID', 'Date', 'Time', 'Time2'], axis=1)
user_pair_scores['User_Pair'] = user_pair_scores['User_ID'].astype('str') + ',' + \
user_pair_scores['User_ID2'].astype('str')
user_pair_scores = user_pair_scores.drop(['User_ID', 'User_ID2'], axis=1)
user_pair_scores.head()

Score      User_Pair
360413      0    15389,99873
357797      0  103754,103755
260743      0    12877,94240
297026      0      2800,3353
445572      0    15466,43153

In [31]:
user_pair_meet_count = user_pair_scores.groupby('User_Pair', sort=False)['User_Pair'].count().\
reset_index(name='Meet_Count')

user_pair_avg_sec = user_pair_scores.groupby('User_Pair', sort=False)['Score'].mean().\
reset_index(name='Average_Seconds')

user_pair_data = pd.concat([user_pair_meet_count, user_pair_avg_sec], axis=1)

In [32]:
user_pair_data.head()

User_Pair  Meet_Count      User_Pair  Average_Seconds
0    15389,99873          34    15389,99873      1651.735294
1  103754,103755          60  103754,103755      4637.683333
2    12877,94240          32    12877,94240      2735.687500
3      2800,3353          14      2800,3353       311.785714
4    15466,43153          11    15466,43153       327.181818

In [33]:
user_pair_data['Score'] = user_pair_data['Average_Seconds'] / user_pair_data['Meet_Count']
user_pair_data = user_pair_data.sort_values('Score')
user_pair_data.head(25)

User_Pair  Meet_Count      User_Pair  Average_Seconds     Score
288      1859,57786          47     1859,57786         1.872340  0.039837
272        8444,100          64       8444,100         3.171875  0.049561
277    40090,132961         181   40090,132961         9.834254  0.054333
485     57786,57785          19    57786,57785         2.578947  0.135734
502      1545,22532          25     1545,22532         3.560000  0.142400
296       2371,2367          13      2371,2367         1.923077  0.147929
282     12395,13105          54    12395,13105         9.685185  0.179355
262       8673,8674           7      8673,8674         1.285714  0.183673
316      6736,79334           8     6736,79334         1.625000  0.203125
666      1119,44897          11     1119,44897         2.727273  0.247934
511     20094,46726          10    20094,46726         3.000000  0.300000
436     57751,57752           6    57751,57752         1.833333  0.305556
727     11221,38293           7    11221,38293         2.142857  0.306122
306    26899,119429           5   26899,119429         4.200000  0.840000
312       2745,4926           2      2745,4926         2.000000  1.000000
155      8797,10759           3     8797,10759         3.666667  1.222222
101      8048,11143           2     8048,11143         2.500000  1.250000
861   101935,165787           2  101935,165787         2.500000  1.250000
248    36781,128912          14   36781,128912        19.500000  1.392857
1345       3379,605           9       3379,605        15.222222  1.691358
21       1051,10410           2     1051,10410         3.500000  1.750000
889         850,849           5        850,849         9.600000  1.920000
380        5177,486           4       5177,486         7.750000  1.937500
1294   66789,155071           2   66789,155071         4.000000  2.000000
281   185288,140477           2  185288,140477         4.500000  2.250000

In [34]:
user_pair_data.loc[user_pair_data['Meet_Count'].idxmax()]

User_Pair          40090,132961
Meet_Count                  181
User_Pair          40090,132961
Average_Seconds         9.83425
Score                 0.0543329
Name: 277, dtype: object

#### The above user pair are obviously friends since they have met a total of 181 times, arriving 9.83 seconds apart on average. Hence they are the highest ranking Friend-Pair based on frequency of meetups.

In [43]:
user_pair_data[user_pair_data['Average_Seconds'].between(180, 420, inclusive=True)]

User_Pair  Meet_Count      User_Pair  Average_Seconds       Score
2412    18813,107599          76   18813,107599       409.868421    5.393006
31     124862,124863          75  124862,124863       419.520000    5.593600
1318       4896,3455          47      4896,3455       268.680851    5.716614
1705       2197,1546          58      2197,1546       337.155172    5.813020
25      27391,119902          42   27391,119902       341.166667    8.123016
818      29488,29488          24    29488,29488       202.541667    8.439236
821       8706,39034          35     8706,39034       344.257143    9.835918
355     113606,22100          19   113606,22100       200.052632   10.529086
202      10282,10283          37    10282,10283       414.486486   11.202337
410    179441,119035          16  179441,119035       182.875000   11.429688
1396       247,22341          18      247,22341       222.111111   12.339506
2530      3554,67714          33     3554,67714       413.696970   12.536272
1615     10440,10439          30    10440,10439       377.066667   12.568889
319     42636,136149          24   42636,136149       301.833333   12.576389
557      32228,69136          25    32228,69136       317.360000   12.694400
1463       576,36259          25      576,36259       317.520000   12.700800
1195      73178,4705          19     73178,4705       242.947368   12.786704
2553     38520,28496          14    38520,28496       187.000000   13.357143
1935     10412,88505          31    10412,88505       415.258065   13.395421
3481      1129,45003          24     1129,45003       335.833333   13.993056
1874     10955,59992          13    10955,59992       181.923077   13.994083
838      49752,35673          29    49752,35673       406.310345   14.010702
4106   178809,178808          19  178809,178808       272.473684   14.340720
1238      5993,35174          19     5993,35174       272.526316   14.343490
441      60607,11101          14    60607,11101       200.857143   14.346939
1174   111138,111139          22  111138,111139       316.000000   14.363636
2406    108425,19224          14   108425,19224       202.500000   14.464286
32       16236,52900          24    16236,52900       348.583333   14.524306
2408   122520,122521          24  122520,122521       357.833333   14.909722
3262        1441,661          16       1441,661       238.562500   14.910156
...              ...         ...            ...              ...         ...
8035   110560,110562           2  110560,110562       410.000000  205.000000
14490   104539,88247           2   104539,88247       410.000000  205.000000
15236        557,559           2        557,559       410.000000  205.000000
12763    25119,74710           2    25119,74710       410.000000  205.000000
10945  149500,187666           2  149500,187666       410.000000  205.000000
15475    12995,12995           2    12995,12995       410.500000  205.250000
12872   108450,19235           2   108450,19235       410.500000  205.250000
10798     2514,32453           2     2514,32453       410.500000  205.250000
3598    154204,84194           2   154204,84194       410.500000  205.250000
14893   118763,26281           2   118763,26281       411.000000  205.500000
15232  104099,104099           2  104099,104099       412.000000  206.000000
13813    10117,10962           2    10117,10962       412.500000  206.250000
15631    41651,41477           2    41651,41477       412.500000  206.250000
12468   120929,58853           2   120929,58853       413.500000  206.750000
10131      14799,211           2      14799,211       414.000000  207.000000
11660      693,39485           2      693,39485       414.000000  207.000000
1138     66635,66636           2    66635,66636       414.000000  207.000000
4431   118869,118872           2  118869,118872       415.000000  207.500000
6759      32165,3018           2     32165,3018       415.500000  207.750000
10025   93475,167723           2   93475,167723       416.500000  208.250000
9588   

### Conclusion:
#### 1. User IDs 40090 & 132961 are clearly the highest scoring Friend-Pair based on the frequency of their meetings (181 times) and their arrival time (9.834 seconds apart) at the same location.
#### 2. User IDs 18813 & 107599 are the winning Non-Friend-Pair based on the frequency (76 times) of arriving at the same location 409.868 seconds (6.831 minutes) apart, which is plausible because a stalker would arrive a few minutes later (but not too much later, I set the range to be 3-7 minutes apart) so as not to arouse suspicion.
#### 3. The scoring system devised was not sophisticated enough to generate a universal score for both criteria. However by manually querying the final dataframe, one could interpret the findings (depending once again on how the Friend and Non-Friend pair is defined).